Doc/c-api/lifecycle.rst

```diff
@@ -0,0 +1,236 @@
+   * When the garbage collector discovers a :term:`cyclic isolate`, it
+     finalizes one of the objects in the group by calling its
+     :c:member:`~PyTypeObject.tp_finalize` function.  This repeats until the
+     cyclic isolate doesn't exist or all of the objects have been finalized.
+   * :c:member:`~PyTypeObject.tp_finalize` can resurrect the object by adding a
+     reference from outside the :term:`cyclic isolate`.  The new reference
+     causes the group of objects to no longer form a cyclic isolate (the
+     reference cycle may still exist; if it does the objects are no longer
+     isolated).
+   * When the garbage collector discovers a :term:`cyclic isolate` and all of
+     the objects in the group have already been finalized, the garbage
+     collector clears one or more of the uncleared objects in the group
+     (possibly concurrently, but with the :term:`GIL` held :ref:`unless
+     disabled <whatsnew313-free-threaded-cpython>`) by calling each's
+     :c:member:`~PyTypeObject.tp_clear` function.  This repeats as long as the
+     cyclic isolate still exists and not all of the objects have been cleared.
+
+
+Cyclic Isolate Destruction
+--------------------------
+
+Listed below are the stages of life of a hypothetical :term:`cyclic isolate`
+that continues to exist after each member object is finalized or cleared.  It
+is a bug if a cyclic isolate progresses through all of these stages; it should
+vanish once all objects are cleared, if not sooner.  A cyclic isolate can
+vanish either because the reference cycle is broken or because the objects are
+no longer isolated due to finalizer resurrection (see
+:c:member:`~PyTypeObject.tp_finalize`).
+
+0. **Reachable** (pre-cyclic isolate): All objects are in their normal,
+   reachable state.  A reference cycle exists, but an external reference means
+   the objects are not yet isolated.
+#. **Unreachable but consistent:** The final reference from outside the group
+   of objects has been removed, causing the objects to become isolated (thus a
+   cyclic isolate is born).  None of the group's objects have been finalized or
+   cleared yet.
+#. **Mix of finalized and not finalized:** Objects in a cyclic isolate are
+   finalized one at a time, which means that there is a period of time when the
+   cyclic isolate is composed of a mix of finalized and non-finalized objects.
+   Finalization order is unspecified, so it can appear random.  A finalized
+   object must behave in a sane manner when non-finalized objects interact with
+   it, and a non-finalized object must be able to tolerate the finalization of
+   an arbitrary subset of its referents.
+#. **All finalized:** All objects in a cyclic isolate are finalized before any
+   of them are cleared.
+#. **Mix of finalized and cleared:** The objects can be cleared serially or
+   concurrently (but with the :term:`GIL` held :ref:`unless disabled
+   <whatsnew313-free-threaded-cpython>`); either way, some will finish before
```

Comment:
Again, saying "GIL might not be held" is probably misleading. But anyways, user won't really notice or care if they're getting cleared concurrently or not.
```suggestion
   concurrently; either way, some will finish before
```

Reply:
They will care if they do any inter-object calls that might require locking so I left the GIL comment there, but removed the reference to free threading.